### PR TITLE
Updates UserProficiency table incrementally.

### PR DESCRIPTION
**Pull request by ashercodeorg**
Note that this PR increases the number of DB reads and (possibly) writes associated with a user finishing a level. It may or may not make sense to store client-side whether the user has requested a hint, eliminating a DB read some of the time.

Note that a script will need to be written and run to back-calculate UserProficiency for existing UserLevel rows.

### CODE REVIEW

**Comment by ashercodeorg** [comment]
CC @philbogle for scale reasons.

**Comment by philbogle** [comment]
I think this won't hurt HOC scale, assuming we turn off progress tracking for high scale levels  as planned during that week.

**Comment by Hamms** [comment]
Particularly if we put it behind a gatekeeper flag. Can we go ahead and do that now?

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -800,6 +800,38 @@ def User.track_script_progress(user_id, script_id)
...
 
+  # If appropriate, increases the level counts for the concept-difficulties
+  # associated with the completed level.
+  def User.track_proficiency(user_id, script_id, level_id)
+    level_concept_difficulty = LevelConceptDifficulty.where(level_id: level_id).first
+    if !level_concept_difficulty
+      return
+    end
+
+    # Exit early if a hint was used, as no proficiency is gained in this case.
```
Comment (Hamms) [inline dashboard/app/models/user.rb]: It seems a little bit odd that the calling function is responsible for checking if perfection has been achieved, but this is responsible for making sure it's REALLY been achieved. What do you think about rolling this check into `new_level_perfected`?

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -800,6 +800,38 @@ def User.track_script_progress(user_id, script_id)
...
+
+    retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
+      user_proficiency = UserProficiency.where(user_id: user_id).first_or_create!
+      user_proficiency.last_progress_at = Time.now
+
+      ConceptDifficulties::CONCEPTS.each do |concept|
+        if level_concept_difficulty.send(concept)
+          # TODO(asher): If possible, simplify this to use direct assignment via
+          # incrementing rather than going through setting attributes.
+          user_proficiency.attributes = {"#{concept}_d#{level_concept_difficulty.send(concept)}_count" => user_proficiency.send("#{concept}_d#{level_concept_difficulty.send(concept)}_count") + 1}
```
Comment (Hamms) [inline dashboard/app/models/user.rb]: this update bit requires enough knowledge about the details of user_proficiency that it seems like it should really go in a method on the model.

Comment (Hamms) [inline dashboard/app/models/user.rb]: re: the TODO. Would something like this work?

``` ruby
field = "#{concept}_d#{level_concept_difficulty.send(concept)}_count"
user_proficiency.send("#{field}=", user_proficiency.send(field) + 1)
```

**Comment by Hamms** [comment]
first look done!

File: dashboard/app/models/user.rb (hunk 1 above)
Comment (ashercodeorg) [inline dashboard/app/models/user.rb]: Done (good idea).

**Comment by ashercodeorg** [comment]
It shouldn't be necessary to put this behind a gatekeeper flag? Doesn't this get disabled automatically (via the same apparatus as) when progress tracking gets disabled?

IMO, if we are tracking user progress, we should also be tracking user proficiency since the aim is to surface user proficiency.

**Comment by ashercodeorg** [comment]
Assuming all tests pass, this should be ready for another look.

**Comment by Hamms** [comment]
Oh, of course, that makes sense.

This looks great, but would you hold off on merging it in until I'm able to get the concept difficulty data into the remaining `blockly` levels?

**Comment by Hamms** [comment]
Oh, this should be good to go; all level_concept_difficulties are populated on production

**Comment by ashercodeorg** [comment]
Roger, I'll submit after confirming tests are still passing after changing the requirement from a score of 30 to a score of 100, as discussed.

**Pull request merged**